docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,316 @@
+  of an experimental flag.
+
+- The yaml processor libraries will be located in a `contrib` subfolder. For
+  example, it could be located within `cluster-api/cmd/clusterctl`.
+
+- Another method of defining a yaml processor instead of using its library
+  would be to shell out to the templating engine binary/executable.
+  The advantage of this is that the templating engine is not a dependency within
+  the clusterctl codebase.
+  The disadvantage is that the user needs to ensure the binary is installed
+  separately which disrupts the UX of clusterctl. If each infrastructure
+  provider provides different template styles, the user would need to install
+  multiple templating tool binaries.
+### Risks and Mitigations
+- Support path for supporting extended templating tools such as ytt within
+  clusterctl. That is, if there are templating issues, who owns them and how
+  do we address them. This is for each templating engine that is included in
+  clusterctl.
+  However since we’ve decided to put extended templating
+  libraries within a **contrib** folder, it will be the responsibility of the
+  contributor and not the project maintainers to support the templating
+libraries so the issue of support should be solved with this contract.
+
+- Each provider may want to implement their own templating standard. This
+  could affect the UX of clusterctl when automatically pulling workload
+  cluster templates from various providers.
+
+
+### Constraints
+- Currently, clusterctl relies on the conformance of file name conventions
+  such as `infrastructure-components.yaml` and
+  `cluster-template-<flavor>.yaml`. Other templating tools might require other
+  conventions to be defined and followed to allow the same “day 1” experience.
+- Some templating tools will require multiple files to be defined rather than
+  a single yaml file. These artifacts will need to be “grouped” together to
+  support current retrieval mechanisms. Currently, `clusterctl config cluster`
+  retrieves templates from multiple sources such as ConfigMaps within a
+  cluster, URL, Github Repository, Local Repository and even the overrides
+  directory. To ensure compatibility, we’ll need to establish a compression
+  format like tar, gzip or zip.
+
+## Alternatives
+- An alternative to allowing clusterctl handle templates would be to just use
+  another tool upstream to generate the various templates with variables
+  defined. Then use clusterctl to consume that template and use the default
+  variable substitution templating solution.
+
+
+## Upgrade Strategy
+
+
+Any individual template engine libraries would be upgraded via go modules
+preferably using semver standards.
+
+
+## Additional Details
+
+
+Currently, cluster-api-provider-azure is using kustomize to author their
+cluster template flavors. See [issue 540] for more details. Maybe we can
+collect input from them to understand their strategy if they need to
+manage their template variations.
```

Comment:
Risk/Mitigation?

Reply:
I don't see this as much of a risk but rather another source of information. I can reach out to @CecileRobertMichon and have a quick discussion about their use case.

Reply:
@wfernandes did you still want to sync about this before the proposal merges?

Reply:
@CecileRobertMichon Sure let's do a zoom call sometime today if you're available and we can sync up 🙂